eth_enr/sqlite3_db.py

```
@@ -0,0 +1,334 @@
+    if record_row is None:
+        raise RecordNotFound(f"No record found: node_id={node_id.hex()}")
+    field_rows = conn.execute(FIELD_GET_QUERY, (node_id, record_row[1])).fetchall()
+
+    fields = tuple(Field.from_row(row) for row in field_rows)
+    record = Record.from_row(row=record_row, fields=fields)
+    return record
+
+
+DELETE_RECORD_QUERY = """DELETE FROM record WHERE record.node_id = ?"""
+DELETE_FIELD_QUERY = """DELETE FROM field WHERE field.node_id = ?"""
+
+
+def delete_record(conn: sqlite3.Connection, node_id: NodeID) -> int:
+    with conn:
+        cursor = conn.execute(DELETE_RECORD_QUERY, (node_id,))
+        conn.execute(DELETE_FIELD_QUERY, (node_id,))
+    return cursor.rowcount  # type: ignore
+
+
+BASE_QUERY = """SELECT
+    record.node_id AS record_node_id,
+    record.sequence_number AS record_sequence_number,
+    record.signature AS record_signature,
+    record.created_at AS record_created_at
+    FROM record
+    INNER JOIN (
+        SELECT
+            record.node_id,
+            record.sequence_number,
+            MAX(record.sequence_number)
+        FROM record
+        GROUP BY record.node_id
+    ) latest_record
+        ON
+            record.node_id == latest_record.node_id AND
+            record.sequence_number == latest_record.sequence_number
+    JOIN field
+        ON
+            record.node_id = field.node_id AND
+            record.sequence_number = field.sequence_number
+    {where_statements}
+    GROUP BY record.node_id
+"""
+
+
+EXISTS_CLAUSE = """EXISTS (
+    SELECT 1
+    FROM field
+    WHERE
+        record.node_id = field.node_id
+        AND record.sequence_number = field.sequence_number AND field."key" = ?
+)
+"""
+
+
+def query_records(
+    conn: sqlite3.Connection, required_keys: Collection[bytes] = ()
+) -> Iterable[Record]:
+    num_required_keys = len(required_keys)
+
+    if num_required_keys == 0:
+        query = BASE_QUERY.format(where_statements="")
```

I wonder if it makes sense to allow this? It'd basically be returning all items, right? Wouldn't it be better to raise an error here and have a separate API that returns all records?

---

I think that'd be better as it'd help us catch mistakes leading to calling `.query()` with an empty list of constraints, but also because using `BASE_QUERY` to fetch all entries is probably less efficient than a bare `select`

---

I'm inclined to allow it.  You're correct about the efficiency bit but I'm not too concerned about it and it is something we can address later if it is actually problematic (opening an issue to track it).